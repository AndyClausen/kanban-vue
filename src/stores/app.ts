// Utilities
import { defineStore } from "pinia";

export interface Column {
  id: number;
  title: string;
  cards: Card[];
}

export interface Card {
  id: number;
  title: string;
  description: string;
}

export const useAppStore = defineStore("app", {
  state: () => ({
    columns: [
      {
        id: 1,
        title: "To Do",
        cards: [
          {
            id: 1,
            title: "Buy groceries",
            description: "Milk, eggs, bread, and butter",
          },
          {
            id: 2,
            title: "Pay Rent",
            description: "Monthly rent for a small apartment",
          },
        ],
      },
      {
        id: 2,
        title: "Doing",
        cards: [
          {
            id: 3,
            title: "Clean the house",
            description: "Dust, vacuum, sweep, mop",
          },
          {
            id: 4,
            title: "Mow the lawn",
            description: "Trim the grass, weed the flowers",
          },
        ],
      },
      {
        id: 3,
        title: "Done",
        cards: [
          {
            id: 5,
            title: "Prepare for the weekend",
            description: "Pack for the trip, buy tickets",
          },
          {
            id: 6,
            title: "Watch the sunset",
            description: "Relax, unwind, and enjoy the beauty",
          },
        ],
      },
    ],
  }),
  actions: {
    addColumn() {
      this.columns.push({
        id: this.columns.length + 1,
        title: "",
        cards: [],
      });
    },
    addCard(columnId: number) {
      const col = this.columns.find((column) => column.id === columnId);
      if (!col) {
        return;
      }
      col.cards.push({
        id: col.cards.length + 1,
        title: "",
        description: "",
      });
    },
  },
});
