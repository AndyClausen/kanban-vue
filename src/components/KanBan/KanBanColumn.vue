<template>
  <v-card :color="color" width="344">
    <v-card-title :class="{ 'font-italic': !column.title }">{{
      column.title || "Untitled"
    }}</v-card-title>
    <v-card-text>
      <v-container>
        <draggable :list="column.cards" itemKey="id" group="cards">
          <template #item="{ element: card }">
            <v-row>
              <v-col cols="12">
                <KanBanCard :card="card" />
              </v-col>
            </v-row>
          </template>
          <template #footer>
            <v-row>
              <v-col cols="12">
                <v-container
                  class="border-lg rounded-lg border-success d-flex justify-center align-center cursor-pointer"
                  width="100%"
                  height="80px"
                  @click="store.addCard(column.id)"
                >
                  <span class="text-success font-weight-bold text-h2">+</span>
                </v-container>
              </v-col>
            </v-row>
          </template>
        </draggable>
      </v-container>
    </v-card-text>
  </v-card>
</template>

<script setup lang="ts">
import draggable from "vuedraggable";
import { useAppStore, type Column } from "@/stores/app";

const { column } = defineProps<{ column: Column }>();
const store = useAppStore();

const color = "#FFECBE";
</script>
