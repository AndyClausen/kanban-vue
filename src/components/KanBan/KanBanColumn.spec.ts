import { mount } from "@vue/test-utils";
import { describe, it, expect, vi } from "vitest";
import { createTestingPinia } from "@pinia/testing";
import { createVuetify } from "vuetify";
import { components, directives } from "vuetify/dist/vuetify.js";
import KanBanColumn from "./KanBanColumn.vue";
import { useAppStore } from "@/stores/app";

const pinia = createTestingPinia({ createSpy: vi.fn, stubActions: false });
const store = useAppStore(pinia);
const wrapper = mount(KanBanColumn, {
  global: {
    plugins: [createVuetify({ components, directives }), pinia],
  },
  props: {
    column: store.columns[0],
  },
});

describe("KanBanColumn", () => {
  it("renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("has an add card button", () => {
    expect(wrapper.find(".v-container.cursor-pointer").exists()).toBe(true);
  });
  it("adds a card when button is clicked", async () => {
    expect(wrapper.findAll("KanBanCard").length).toBe(2);
    await wrapper.get(".v-container.cursor-pointer").trigger("click");
    expect(wrapper.findAll("KanBanCard").length).toBe(3);
  });
});
