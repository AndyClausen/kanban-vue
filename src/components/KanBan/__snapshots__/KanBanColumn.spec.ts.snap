// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KanBanColumn > renders the component 1`] = `
"<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated" style="background-color: #FFECBE; color: #000; caret-color: #000; width: 344px;">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="--v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-title">To Do</div>
  <div class="v-card-text">
    <div class="v-container v-locale--is-ltr">
      <div>
        <div class="v-row" data-draggable="true">
          <div class="v-col v-col-12">
            <kanbancard card="[object Object]"></kanbancard>
          </div>
        </div>
        <div class="v-row" data-draggable="true">
          <div class="v-col v-col-12">
            <kanbancard card="[object Object]"></kanbancard>
          </div>
        </div>
        <div class="v-row">
          <div class="v-col v-col-12">
            <div class="v-container v-locale--is-ltr border-lg rounded-lg border-success d-flex justify-center align-center cursor-pointer" style="height: 80px; width: 100%;"><span class="text-success font-weight-bold text-h2">+</span></div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
